### DISCUSSION TITLE
Add Hosting "this.CreateBuilder" is not recognized although the features and usings are added?

### DISCUSSION THREAD

**Discussion by DevTKSS** [unoplatform/uno#18904] [Q&A]
Hey,
I did  try to add in a blanc templated uno app Hosting like: https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Tutorials/Hosting/HowTo-HostingSetup.html

Added usings for Microsoft.Extensions.DependencyInjection and Microsoft.Extensions.Hosting and the package References and features in uno features. But something seems to be missing, because the code from the guide results in intellisense telling, CreateBuilder(args) is not recognized... Is there maybe something else that needs to be imported in the project beside this parts mentioned? Replicated exactly what you see in the guide, also the IHost above this void.

Thanks in advance

**Comment by kazo0** [comment]
@DevTKSS would you be able to provide your repro project as a .zip here? I am walking through the guide as well to make a blank project but it might help to see how yours is configured just in case there's a difference somewhere that I can spot

Thanks!

**Reply by DevTKSS** [reply]
![image](https://github.com/user-attachments/assets/29ef0cd5-57be-41a7-9d8c-8d1247905a93)
on mine the IHost is not recognized, so I doubt that is actually helping you...
https://github.com/DevTKSS/TheCatApiClient.git
That is why I did create that question, to spot what is wrong. I did add the tag into the unoFeatures like decribed, but did not do the trick. started vs 2022 new, try to build, but nothing. all Intellisense is telling, I should add Microsoft.Extensions.Hosting, but uno docs did explicitly tell that Uno is the namespace to use, since they did extend the microsoft one :)

**Reply by kazo0** [reply]
You will still need to add a using for the `Microsoft.Extensions.Hosting` namespace. So, depending on how the project was created, it may be added for you already in the `GlobalUsings.cs`. For example, if I went through the Visual Studio Uno Wizard and selected Dependency Injection:

<img width="1011" alt="image" src="https://github.com/user-attachments/assets/ab3936ad-ed57-4321-9bcc-b601ae3e1b9f">

I end up with this in my `GlobalUsings.cs`:

![image](https://github.com/user-attachments/assets/d25fb646-30d3-44ee-a572-c5a8f55f251e)

And the `App.xaml.cs`:

<img width="1257" alt="image" src="https://github.com/user-attachments/assets/e81cc82e-c6c4-41ed-9e6a-7118949876c6">



I'd like to find out where in the guide we should add some more explanations surrounding this. When working through the guide, did you follow the steps within the Warning here?

![image](https://github.com/user-attachments/assets/ec23ea27-f8d7-4043-94e7-738496e9915e)

If so, it should have added the `using Microsoft.Extensions.Hosting;` to the `GlobalUsings.cs`. Otherwise, if you manually installed the Hosting extension NuGet package, it would make sense that this `using` would need to be added manually.

**Reply by DevTKSS** [reply]
I got totally confused by this informations and warnings, because they are sending me from A>B>A>B in infinity... In the "add in exising project" guide there is just not providing any more information other than in the other guide. The Hosting.Winui one you shooted there is in my app dependencys and not doing anything. so from they telling to add the feature tag and my proj did add the dependency there is no where mentioned to add a using to microsoft, isnt it? so I would assume that I should not do that for avoiding problems in namespaces. But yes, as you say, generated from creation with hosting/DI, it is using that. maybe that should be more detailed in guide.
